libminifi/src/io/ClientSocket.cpp

```diff
@@ -0,0 +1,621 @@
+      else logger_->log_info("Bind to interface %s", local_network_interface_.getInterface());
+    }
+#endif /* !WIN32 */
+    sockaddr_in sa_loc{};
+    memset(&sa_loc, 0x00, sizeof(sa_loc));
+    sa_loc.sin_family = AF_INET;
+    sa_loc.sin_port = htons(port_);
+    // use any address if you are connecting to the local machine for testing
+    // otherwise we must use the requested hostname
+    if (IsNullOrEmpty(requested_hostname_) || requested_hostname_ == "localhost") {
+      sa_loc.sin_addr.s_addr = htonl(is_loopback_only_ ? INADDR_LOOPBACK : INADDR_ANY);
+    } else {
+#ifdef WIN32
+      sa_loc.sin_addr.s_addr = addr.s_addr;
+    }
+    if (connect(socket_file_descriptor_, reinterpret_cast<const sockaddr*>(&sa_loc), sizeof(sockaddr_in)) == SOCKET_ERROR) {
+      int err = WSAGetLastError();
+      if (err == WSAEADDRNOTAVAIL) {
+        logger_->log_error("invalid or unknown IP");
+      } else if (err == WSAECONNREFUSED) {
+        logger_->log_error("Connection refused");
+      } else {
+        logger_->log_error("Unknown error");
+      }
+#else
+      sa_loc.sin_addr.s_addr = addr;
+    }
+    if (connect(socket_file_descriptor_, reinterpret_cast<const sockaddr *>(&sa_loc), sizeof(sockaddr_in)) < 0) {
+#endif /* WIN32 */
+      closeStream();
+      return -1;
+    }
+  }
+
+  // add the listener to the total set
+  FD_SET(socket_file_descriptor_, &total_list_);
+  socket_max_ = socket_file_descriptor_;
+  logger_->log_debug("Created connection with file descriptor %d", socket_file_descriptor_);
+  return 0;
+}
+
+int16_t Socket::initialize() {
+  addrinfo hints{};
+  memset(&hints, 0, sizeof hints);  // make sure the struct is empty
+  hints.ai_family = AF_UNSPEC;
+  hints.ai_socktype = SOCK_STREAM;
+  hints.ai_flags = AI_CANONNAME;
+  if (listeners_ > 0 && !is_loopback_only_)
+    hints.ai_flags = AI_PASSIVE;
+  hints.ai_protocol = 0; /* any protocol */
+
+  const char* const gai_node = [this]() -> const char* {
+    if (is_loopback_only_) return "localhost";
+    if (!is_loopback_only_ && listeners_ > 0) return nullptr;  // all non-localhost server sockets listen on wildcard address
+    if (!requested_hostname_.empty()) return requested_hostname_.c_str();
+    return nullptr;
+  }();
+  const auto gai_service = std::to_string(port_);
+  addrinfo* getaddrinfo_result = nullptr;
+  const int errcode = getaddrinfo(gai_node, gai_service.c_str(), &hints, &getaddrinfo_result);
+  const std::unique_ptr<addrinfo, util::addrinfo_deleter> addr_info{ getaddrinfo_result };
+  getaddrinfo_result = nullptr;
+  if (errcode != 0) {
+    logger_->log_error("getaddrinfo: %s", get_last_getaddrinfo_err_str(errcode));
+    return -1;
+  }
+  socket_file_descriptor_ = INVALID_SOCKET;
+
+  // AI_CANONNAME always sets ai_canonname of the first addrinfo structure
+  canonical_hostname_ = !IsNullOrEmpty(addr_info->ai_canonname) ? addr_info->ai_canonname : requested_hostname_;
```

Comment:
Behavior change: In the old code, we did a name lookup for the canonical name even for server sockets. This is no longer the case since now there is only one name lookup instead of two and we specify a wildcard node to `getaddrinfo` so that we get a result with wildcard addresses back. This means that the name resolution is no longer performed and we no longer set the `canonical_hostname_` member. As a stopgap, I'm assigning `requested_hostname_` to `canonical_hostname_`, but if this change in behavior is not acceptable, then we need to do a second name lookup in those cases anyway.